### ISSUE TITLE
data_pre.py

### ISSUE THREAD

**Issue by milesbrown1102** [milesbrown1102/milebrown1102.github.io#1]
import pandas as pd
import numpy as np
from sklearn.neighbors import NearestNeighbors
from sklearn.impute import KNNImputer
from scipy.stats import pointbiserialr

def load_data(file_path):
    """Load the dataset and return the DataFrame."""
    df = pd.read_excel(file_path, sheet_name="data", engine="openpyxl")
    return df

def create_missing_values(df, missing_percentage):
    """
    Introduce missing values into all columns except the first one.
    Missing values are added uniformly across the dataset.
    """
    df_missing = df.copy()
    np.random.seed(35)  # For reproducibility
    for col in df_missing.columns[1:]:
        mask = np.random.rand(len(df_missing)) < (missing_percentage / 100)
        df_missing.loc[mask, col] = np.nan
    return df_missing

def split_data(df):
    """
    Split the data based on the first column (Bankruptcy indicator).
    Returns three DataFrames:
      - bankrupt_df: rows where the indicator equals 1,
      - not_bankrupt_df: rows where the indicator equals 0,
      - missing_bankrupt_df: rows where the indicator is missing.
    """
    bankrupt_df = df[df.iloc[:, 0] == 1]
    not_bankrupt_df = df[df.iloc[:, 0] == 0]
    missing_bankrupt_df = df[df.iloc[:, 0].isna()]
    return bankrupt_df, not_bankrupt_df, missing_bankrupt_df

def knn_impute(df, n_neighbors):
    """
    Apply KNN imputation using sklearn's KNNImputer.
    This method fills missing values based on the closest feature distances.
    """
    df_imputed = df.copy()
    imputer = KNNImputer(n_neighbors=n_neighbors)
    df_imputed.iloc[:, 1:] = imputer.fit_transform(df_imputed.iloc[:, 1:])  # Exclude the first column
    return df_imputed

def impute_data(df):
    """
    Generate multiple imputation versions for the given DataFrame (assumed to be one subset).
    Imputation is performed independently for each column (excluding the bankruptcy indicator).
    Returns a dictionary with:
      - "Mean": Missing values replaced with the column mean.
      - "Median": Missing values replaced with the column median.
      - "KNN-5": KNN imputation using 5 neighbors.
      - "KNN-15": KNN imputation using 15 neighbors.
      - "KNN-30": KNN imputation using 30 neighbors.
    """
    # Mean imputation
    df_mean = df.copy()
    for col in df_mean.columns[1:]:
        df_mean[col] = df_mean[col].fillna(df_mean[col].mean())

    # Median imputation
    df_median = df.copy()
    for col in df_median.columns[1:]:
        df_median[col] = df_median[col].fillna(df_median[col].median())

    # KNN imputation with different neighbor counts
    df_knn_5 = knn_impute(df, 5)
    df_knn_15 = knn_impute(df, 15)
    df_knn_30 = knn_impute(df, 30)

    return {
        "Mean": df_mean,
        "Median": df_median,
        "KNN-5": df_knn_5,
        "KNN-15": df_knn_15,
        "KNN-30": df_knn_30,
    }

def save_to_excel(imputed_data, original_df, file_path):
    """
    Save all imputed datasets into an Excel file.
    Each imputation method is saved to a separate sheet for:
      - Bankrupt (indicator = 1)
      - Non-Bankrupt (indicator = 0)
      - (Rows with missing bankruptcy indicator, if any)
    Also includes the original values for comparison.
    """
    with pd.ExcelWriter(file_path, engine="openpyxl") as writer:
        for imputation_method, df in imputed_data.items():
            bankrupt_df, not_bankrupt_df, missing_bankrupt_df = split_data(df)
            original_df_with_imputed = original_df.copy()
            for col in original_df.columns[1:]:
                original_df_with_imputed[f"Imputed_{imputation_method}_{col}"] = df[col]
            if not bankrupt_df.empty:
                bankrupt_df.to_excel(writer, sheet_name=f"Bankrupt_{imputation_method}", index=False)
            if not not_bankrupt_df.empty:
                not_bankrupt_df.to_excel(writer, sheet_name=f"Not_Bankrupt_{imputation_method}", index=False)
            if not missing_bankrupt_df.empty:
                missing_bankrupt_df.to_excel(writer, sheet_name=f"Missing_Bankrupt_{imputation_method}", index=False)
            original_df_with_imputed.to_excel(writer, sheet_name=f"Original_and_Imputed_{imputation_method}", index=False)
    print(f"Imputed data saved to {file_path}")

def calculate_closeness(original_df, imputed_data, missing_indices):
    """
    For each missing cell in original_df (as indicated by missing_indices), compare the
    imputed values from each method to the original value,
    and count how many times each method is closest to the original.
    Returns a dictionary with counts for each imputation method.
    """
    closest_method_counts = {method: 0 for method in imputed_data}
    for col in original_df.columns[1:]:
        missing_mask = missing_indices[col]
        for i in original_df[missing_mask].index:
            original_value = original_df.loc[i, col]
            imputed_values = {method: imputed_data[method].loc[i, col] for method in imputed_data}
            differences = {method: abs(original_value - imputed_val) for method, imputed_val in imputed_values.items()}
            closest_method = min(differences, key=differences.get)
            closest_method_counts[closest_method] += 1
    return closest_method_counts

def calculate_correlations(original_df, imputed_data, missing_indices):
    """
    For each imputation method, compute the Pearson correlation coefficient between the original values and
    the imputed values for the cells that were artificially made missing.
    Returns a dictionary with the average correlation (averaged over columns with at least two missing values)
    for each imputation method.
    """
    correlation_results = {}

    # Loop through each imputation method
    for method, df_imputed in imputed_data.items():
        correlations = []

        # Loop through each column to calculate correlations
        for col in original_df.columns[1:]:  # Skipping the first column (likely an ID column)
            mask = missing_indices[col]

            if mask.sum() >= 2:  # We only want columns with at least two missing values
                orig_vals = original_df.loc[mask, col]
                imp_vals = df_imputed.loc[mask, col]

                # Check if either original or imputed values are constant (std == 0)
                if np.std(orig_vals) == 0 or np.std(imp_vals) == 0:
                    corr = 1.0  # If there's no variance, treat correlation as perfect
                else:
                    try:
                        # Compute the Pearson correlation
                        corr = np.corrcoef(orig_vals, imp_vals)[0, 1]
                    except Exception as e:
                        print(f"Error calculating correlation for {method} and column {col}: {e}")
                        corr = np.nan  # Set correlation as NaN if there's an error

                # Only append valid correlations (no NaNs)
                if not np.isnan(corr):
                    correlations.append(corr)

        # Store the average correlation for the method, ensuring it's a float (NaN is fine)
        if correlations:
            correlation_results[method] = np.mean(correlations)
        else:
            correlation_results[method] = np.nan  # No valid correlations, set to NaN

    return correlation_results

def filter_correlations_by_threshold(original_df, imputed_data, missing_indices, threshold):
    """
    For each column (excluding the bankruptcy indicator) and for each imputation method,
    compute the Pearson correlation coefficient between the original and imputed values
    (only considering cells where missing values were introduced).
    Returns a list of tuples (column_name, imputation_method, correlation_value) for
    those cases where the correlation meets or exceeds the threshold.
    """
    filtered_results = []
    for col in original_df.columns[1:]:
        mask = missing_indices[col]
        if mask.sum() >= 2:
            orig_vals = original_df.loc[mask, col]
            for method, df_imputed in imputed_data.items():
                imputed_vals = df_imputed.loc[mask, col]
                if np.std(orig_vals) == 0 or np.std(imputed_vals) == 0:
                    corr = 1.0
                else:
                    corr = np.corrcoef(orig_vals, imputed_vals)[0, 1]
                if corr >= threshold:
                    filtered_results.append((col, method, corr))
    return filtered_results

def point_biserial_correlation_heatmap(df, binary_column, p_value_threshold=0.05, correlation_threshold=0.2):
    """
    Compute the point-biserial correlation coefficient between a binary target column
    and each continuous feature in the dataframe, and store all correlations.

    Returns a dictionary of all features with their correlations and p-values.
    """
    correlation_results = {}

    for col in df.columns:
        if col != binary_column:
            corr, p_value = pointbiserialr(df[binary_column], df[col])

            # Apply correlation and p-value thresholds
            if abs(corr) > correlation_threshold and p_value < p_value_threshold:
                correlation_results[col] = {"Correlation": corr, "p-value": p_value}

    return correlation_results

def point_biserial_correlation(df, binary_column, p_value_threshold=0.05, correlation_threshold=0.2, top_n=20):
    """
    Compute the point-biserial correlation coefficient between a binary target column
    and each continuous feature in the dataframe, and select the top N features.

    Features must meet:
      - Absolute correlation > correlation_threshold (default: 0.2)
      - p-value < p_value_threshold (default: 0.05)

    Returns a dictionary of the top N selected features sorted by absolute correlation.
    """
    correlation_results = {}

    for col in df.columns:
        if col != binary_column:
            corr, p_value = pointbiserialr(df[binary_column], df[col])

            # Apply correlation and p-value thresholds
            if abs(corr) > correlation_threshold and p_value < p_value_threshold:
                correlation_results[col] = {"Correlation": corr, "p-value": p_value}

    # Sort by absolute correlation in descending order
    sorted_features = sorted(correlation_results.items(), key=lambda x: abs(x[1]["Correlation"]), reverse=True)

    # Select the top N features
    selected_features = dict(sorted_features[:top_n])

    return selected_features

def filter_selected_features(original_df, selected_features, output_file_path):
    """
    Filter the original DataFrame to only include the selected features and save to a new Excel file.

    Arguments:
    - original_df: The original DataFrame with all features.
    - selected_features: A dictionary or list of selected feature columns.
    - output_file_path: The file path where the filtered DataFrame should be saved.
    """
    # If selected_features is a dictionary from point_biserial_correlation, we only need the feature names
    if isinstance(selected_features, dict):
        selected_columns = list(selected_features.keys())
    else:
        selected_columns = selected_features

    # Ensure the target column (Bankruptcy indicator) is included if needed
    selected_columns = ['Bankrupt?'] + selected_columns if 'Bankruptcy' in original_df.columns else selected_columns

    # Filter the DataFrame to include only the selected columns
    filtered_df = original_df[selected_columns]

    # Save the filtered DataFrame to a new Excel file
    with pd.ExcelWriter(output_file_path, engine="openpyxl") as writer:
        filtered_df.to_excel(writer, index=False)

    print(f"Filtered data saved to {output_file_path}")
